src/empire/city.c

```diff
@@ -258,39 +268,18 @@ static int generate_trader(int city_id, empire_city *city)
     if (trade_potential <= 0) {
         return 0;
     }
-    int max_traders = calc_bound(trade_potential / RESOURCES_TO_TRADER_RATIO + 1, 1, 3);
-    int index;
-    if (max_traders == 1) {
-        if (!city->trader_figure_ids[0]) {
-            index = 0;
-        } else {
-            return 0;
-        }
-    } else if (max_traders == 2) {
-        if (!city->trader_figure_ids[0]) {
-            index = 0;
-        } else if (!city->trader_figure_ids[1]) {
-            index = 1;
-        } else {
-            return 0;
-        }
-    } else { // 3
-        if (!city->trader_figure_ids[0]) {
-            index = 0;
-        } else if (!city->trader_figure_ids[1]) {
-            index = 1;
-        } else if (!city->trader_figure_ids[2]) {
-            index = 2;
-        } else {
-            return 0;
+
+    // Find a slot to hold a trader
+    int max_traders = calc_bound(trade_potential / RESOURCES_TO_TRADER_RATIO + 1, 1, MAX_TRADERS);
+    int index = -1;
+    for (int f = 0; f < max_traders; f++) {
```

Comment:
Nitpick: Though we use `f` for the figure pointer, for figure array indexes we do use `i`.